---
title: "PARUS1 PopGen Analysis"
author: "Ellis Lab"
date: "`r Sys.Date()`"
output: html_document
---

### Background
This is the analysis of the host and parasite mapping and variant calling from run2. The first code block is a shell script that I ran as an interactive slurm session. The second is `R` code for processing/plotting the results.

#### Code
The shell script. I copied most of the main results in the script.
```{bash, eval=FALSE}
####################################
## define environmental variables ##
####################################

## programs
PLINK=/home/vaellis/packages/plink
VCFTOOLS=/home/vaellis/packages/vcftools/src/cpp/vcftools

## existing directories and files
START_DIR=/work/vaellis/parus1_swga_methods_final
REF_DIR=${START_DIR}/reference_genomes
REF_H=${REF_DIR}/GCA_030015615.1_cyaCae_Pacbio_Achrs_GRC_genomic.fna
REF_P=${REF_DIR}/WW2_genome_against_SISKIN1.fasta
HOST_VCF_DIR=${START_DIR}/12_host_joint_vcf
PATH_VCF_DIR=${START_DIR}/13_pathogen_joint_vcf
VCF_H=${HOST_VCF_DIR}/host_joint_final.vcf.gz
VCF_P=${PATH_VCF_DIR}/pathogen_joint_final.vcf.gz

## new directories and files
POPGEN_DIR=/work/vaellis/parus1_swga_popgen
POP_HOST_DIR=${POPGEN_DIR}/host
POP_PATH_DIR=${POPGEN_DIR}/pathogen
VCF_H_F=${POP_HOST_DIR}/host_filt_joint_final.vcf.gz
VCF_P_F=${POP_PATH_DIR}/path_filt_joint_final.vcf.gz

mkdir -p ${POPGEN_DIR}
mkdir -p ${POP_HOST_DIR}
mkdir -p ${POP_PATH_DIR}

##################
## Examine VCFs ##
##################

## host
# count SNPs (14276887) and Indels (2344956), MNPs (0), other (0)
bcftools view -v snps ${VCF_H} | grep -v "^#" | wc -l
bcftools view -v indels ${VCF_H} | grep -v "^#" | wc -l
bcftools view -v mnps ${VCF_H} | grep -v "^#" | wc -l
bcftools view -v other ${VCF_H} | grep -v "^#" | wc -l

# examine the variable sites (skip unless interactive slurm job)
# bcftools query -f '%CHROM %POS %REF %ALT\n[%SAMPLE %GT\n]\n' ${VCF_H} | less

# mean (0.911852) and se (0.0004162) of proportion of missing genotypes (individuals) for each variable site
bcftools +fill-tags ${VCF_H}  -- -t 'NMISS=N_MISSING' | \
bcftools query -f '%NMISS\n' | \
awk '{sum+=$1; sumsq+=($1)^2; count++} END {mean=sum/count; variance=(sumsq/count - mean^2); se=sqrt(variance/count); printf "Mean: %.10f\nStandard Error: %.10f\n", mean, se}'

# produce two column table: number of missing genotypes and number of variants
## 0 10712742
## 1 2439489
## 2 1171620
## 3 704459
## 4 481758
## 5 354706
## 6 272229
## 7 213916
## 8 166781

bcftools +fill-tags ${VCF_H}  -- -t 'NMISS=N_MISSING' | \
bcftools query -f '%NMISS\n' | \
awk '{n[$1]++} END {for(i in n) print i,n[i]}' | \
sort -n -b -k 1

# mean (123.1912048) and se (0.03051587) of depth per variant
bcftools query -f '%DP\n' ${VCF_H} | \
awk '{sum+=$1; sumsq+=($1)^2; count++} END {mean=sum/count; variance=(sumsq/count - mean^2); se=sqrt(variance/count); printf "Mean: %.10f\nStandard Error: %.10f\n", mean, se}'

# mean (1012.7786116752) and se (0.4861421366) of quality per variant
bcftools query -f '%QUAL\n' ${VCF_H} | \
awk '{sum+=$1; sumsq+=($1)^2; count++} END {mean=sum/count; variance=(sumsq/count - mean^2); se=sqrt(variance/count); printf "Mean: %.10f\nStandard Error: %.10f\n", mean, se}'


## path
# count SNPs (787509) and Indels (86449), MNPs (0), Other (0)
bcftools view -v snps ${VCF_P} | grep -v "^#" | wc -l
bcftools view -v indels ${VCF_P} | grep -v "^#" | wc -l
bcftools view -v mnps ${VCF_P} | grep -v "^#" | wc -l
bcftools view -v other ${VCF_P} | grep -v "^#" | wc -l


# examine the variable sites (ploidy is correct; skip unless interactive slurm job)
# bcftools query -f '%CHROM %POS %REF %ALT\n[%SAMPLE %GT\n]\n' ${VCF_P} | less

# mean (6.7371) and se (0.00145) of proportion of missing genotypes (individuals) for each variable site
bcftools +fill-tags ${VCF_P}  -- -t 'NMISS=N_MISSING' | \
bcftools query -f '%NMISS\n' | \
awk '{sum+=$1; sumsq+=($1)^2; count++} END {mean=sum/count; variance=(sumsq/count - mean^2); se=sqrt(variance/count); printf "Mean: %.10f\nStandard Error: %.10f\n", mean, se}'

# produce two column table: number of missing genotypes and number of variants
## 0 104
## 1 1044
## 2 5075
## 3 16515
## 4 43518
## 5 88368
## 6 153876
## 7 234177
## 8 330155
bcftools +fill-tags ${VCF_P}  -- -t 'NMISS=N_MISSING' | \
bcftools query -f '%NMISS\n' | \
awk '{n[$1]++} END {for(i in n) print i,n[i]}' | \
sort -n -b -k 1

# mean (14.521946) and se (0.016744317) of depth per variant
bcftools query -f '%DP\n' ${VCF_P} | \
awk '{sum+=$1; sumsq+=($1)^2; count++} END {mean=sum/count; variance=(sumsq/count - mean^2); se=sqrt(variance/count); printf "Mean: %.10f\nStandard Error: %.10f\n", mean, se}'

# mean (436.0150057163) and se (0.6450614085) of quality per variant
bcftools query -f '%QUAL\n' ${VCF_P} | \
awk '{sum+=$1; sumsq+=($1)^2; count++} END {mean=sum/count; variance=(sumsq/count - mean^2); se=sqrt(variance/count); printf "Mean: %.10f\nStandard Error: %.10f\n", mean, se}'


#################
## Filter VCFs ##
#################

## host
# restrict to sites with no more than 3 missing genotypes, minimum depth of 5,
# minimum quality of 30
bcftools +fill-tags ${VCF_H}  -- -t 'NMISS=N_MISSING' | \
bcftools filter -i 'NMISS<=3 && FORMAT/DP>=5 && QUAL>=30' -O z -o ${VCF_H_F}

# final number of SNPs (12806876) and Indels (2118287)
bcftools view -v snps ${VCF_H_F} | grep -v "^#" | wc -l
bcftools view -v indels ${VCF_H_F} | grep -v "^#" | wc -l

# final number of SNPs (320843) and Indels (380293) with >1 alternative allele (i.e., variation among samples)
bcftools view -i 'N_ALT > 1' -v snps ${VCF_H_F} | grep -v "^#" | wc -l
bcftools view -i 'N_ALT > 1' -v indels ${VCF_H_F} | grep -v "^#" | wc -l

# final number of multiallelic sites (603872) (note that there can be variants that are both SNPs and indels so this number is typically less than the sum of SNPs and indels)
bcftools stats  ${VCF_H_F} | grep "SN[[:space:]]\+0[[:space:]]\+number of multiallelic sites:"

## path
# restrict to sites with no more than 3 missing genotypes, minimum depth of 5,
# minimum quality of 30
bcftools +fill-tags ${VCF_P}  -- -t 'NMISS=N_MISSING' | \
bcftools filter -i 'NMISS<=3 && FORMAT/DP>=5 && QUAL>=30' -O z -o ${VCF_P_F}

# final number of SNPs (17192) and Indels (3803)
bcftools view -v snps ${VCF_P_F} | grep -v "^#" | wc -l
bcftools view -v indels ${VCF_P_F} | grep -v "^#" | wc -l

# final number of SNPs (187) and Indels (161) with >1 alternative allele (i.e., variation among samples)
bcftools view -i 'N_ALT > 1' -v snps ${VCF_P_F} | grep -v "^#" | wc -l
bcftools view -i 'N_ALT > 1' -v indels ${VCF_P_F} | grep -v "^#" | wc -l

# final number of multiallelic sites (307) (note that there can be variants that are both SNPs and indels so this number is typically less than the sum of SNPs and indels)
bcftools stats  ${VCF_P_F} | grep "SN[[:space:]]\+0[[:space:]]\+number of multiallelic sites:"

# examine the number of sites that vary among individuals in the population only (slightly different than the previous measures, see: https://www.biostars.org/p/9604906/)
bcftools query -f '[%GT]\n' ${VCF_P_F} | awk '{ gsub(/\./, ""); print }' | grep -Ev '^(.)\1+$' | wc -l # 7068 variable sites
bcftools query -f '[%GT]\n' ${VCF_H_F} | awk '{ gsub(/[./|]/, ""); print }' | grep -Ev '^(.)\1+$' | wc -l # 14512339 variable sites

#########
## PCA ##
#########

## host
cd ${POP_HOST_DIR}

# identify and remove variants in linkage disequilibrium; 10kb window size, 10bp step size, linkage if r^2 > 0.2 [let's take this out since we don't really know where these genes sit]
${PLINK} --vcf ${VCF_H_F} --double-id --allow-extra-chr \
--set-missing-var-ids @:# --indep-pairwise 10 10 0.20 --out host

# run PCA, notice the "var-wts", use this to check SNP loadings
${PLINK} --vcf ${VCF_H_F} --double-id --allow-extra-chr \
--set-missing-var-ids @:# \
--extract host.prune.in \
--make-bed --pca var-wts --out host

## path
cd ${POP_PATH_DIR}

# identify and remove variants in linkage disequilibrium; 10kb window size, 10bp step size, linkage if r^2 > 0.2
${PLINK} --vcf ${VCF_P_F} --double-id --allow-extra-chr \
--set-missing-var-ids @:# --indep-pairwise 10 10 0.20 --out pathogen

# run PCA, notice the "var-wts", use this to check SNP loadings
${PLINK} --vcf ${VCF_P_F} --double-id --allow-extra-chr \
--set-missing-var-ids @:# \
--extract pathogen.prune.in \
--make-bed --pca var-wts --out pathogen

################
## Tajima's D ##
################

## host
cd ${POP_HOST_DIR}

${VCFTOOLS} --gzvcf ${VCF_H_F} --TajimaD 1000 --out host_1kb

# count number of windows calculated
tail -n +2 host_1kb.Tajima.D | wc -l

## path
cd ${POP_PATH_DIR}

${VCFTOOLS} --gzvcf ${VCF_P_F} --haploid --TajimaD 1000 --out path_1kb

# count number of windows calculated
tail -n +2 path_1kb.Tajima.D | wc -l

```

`R` code.
```{r, message=FALSE, warning=FALSE}
## set working directory
setwd("C:/Users/vaellis/OneDrive - University of Delaware - o365/Documents/Ellis_Lab/Research/Malaria/SWGA/PARUS1 Methods Paper/")

## packages
library(tidyverse)
library(ggpubr)
library(patchwork)

#########
## PCA ##
#########

## HOST
# read the data
pca.h <- read_table("parus1_swga_popgen/host/host.eigenvec", col_names = FALSE)
eigenval.h <- scan("parus1_swga_popgen/host/host.eigenval")

# sort out the pca data
# remove nuisance column
pca.h <- pca.h[,-1]
# set names
names(pca.h)[1] <- "ind"
names(pca.h)[2:ncol(pca.h)] <- paste0("PC", 1:(ncol(pca.h)-1))

# clean up the data and add year
pca.h.n <- pca.h %>%
  mutate(ind = sub("_S.+", "", ind)) %>%
  mutate(Year = str_extract(ind, "[0-9]+(?=_?)")) %>%
  mutate(Year = factor(Year, levels = c("96", "08", "21"))) %>%
  mutate(Year = fct_recode(Year, "1996" = "96",
                           "2008" = "08",
                           "2021" = "21"))

# now plot the percent variance explained by each PC axis
pve.h <- data.frame(PC = 1:length(eigenval.h), pve = eigenval.h/sum(eigenval.h)*100)
ggplot(pve.h, aes(PC, pve)) + 
  geom_bar(stat = "identity") + 
  ylab("Percentage variance explained") + 
  scale_x_continuous("PC axes", breaks = 1:9) +
  theme_bw()

# calculate the cumulative sum of the percentage variance explained
cumsum(pve.h$pve)

# now plot the samples in the space of the first two PC axes
a <- ggplot(pca.h.n, aes(PC1, PC2)) + 
  geom_point(aes(color = Year, shape = Year), size = 2) +
  xlab(paste0("PC1 (", signif(pve.h$pve[1], 3),"%)")) + 
  ylab(paste0("PC2 (", signif(pve.h$pve[2], 3),"%)")) +
  ggtitle("Host") +
  theme_bw() +
  theme(legend.position = "top",
        legend.title = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())

# now plot the samples in the space of the PC axes 2 and 3
b <- ggplot(pca.h.n, aes(PC2, PC3)) + 
  geom_point(aes(color = Year, shape = Year), size = 2) +
  xlab(paste0("PC2 (", signif(pve.h$pve[2], 3),"%)")) + 
  ylab(paste0("PC3 (", signif(pve.h$pve[3], 3),"%)")) +
  ggtitle("Host") +
  theme_bw() +
  theme(legend.position = "top",
        legend.title = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())

# join the plots
ggarrange(a, b,
          labels = c("A", "B"),
          ncol = 2)

## PATH
# read the data
pca.p <- read_table("parus1_swga_popgen/pathogen/pathogen.eigenvec", col_names = FALSE)
eigenval.p <- scan("parus1_swga_popgen/pathogen/pathogen.eigenval")

# sort out the pca data
# remove nuisance column
pca.p <- pca.p[,-1]
# set names
names(pca.p)[1] <- "ind"
names(pca.p)[2:ncol(pca.p)] <- paste0("PC", 1:(ncol(pca.p)-1))

# clean up the data and add year
pca.p.n <- pca.p %>%
  mutate(ind = sub("_S.+", "", ind)) %>%
  mutate(Year = str_extract(ind, "[0-9]+(?=_?)")) %>%
  mutate(Year = factor(Year, levels = c("96", "08", "21"))) %>%
  mutate(Year = fct_recode(Year, "1996" = "96",
                           "2008" = "08",
                           "2021" = "21"))

# now plot the percent variance explained by each PC axis
pve.p <- data.frame(PC = 1:length(eigenval.p), pve = eigenval.p/sum(eigenval.p)*100)
ggplot(pve.p, aes(PC, pve)) + 
  geom_bar(stat = "identity") + 
  ylab("Percentage variance explained") + 
  scale_x_continuous("PC axes", breaks = 1:9) +
  theme_bw()

# calculate the cumulative sum of the percentage variance explained
cumsum(pve.p$pve)

# now plot the samples in the space of the first two PC axes
c <- ggplot(pca.p.n, aes(PC1, PC2)) + 
  geom_point(aes(color = Year, shape = Year), size = 2) +
  xlab(paste0("PC1 (", signif(pve.p$pve[1], 3),"%)")) + 
  ylab(paste0("PC2 (", signif(pve.p$pve[2], 3),"%)")) +
  ggtitle("Parasite") +
  theme_bw() +
  theme(legend.position = "top",
        legend.title = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())

# now plot the samples in the space of the PC axes 2 and 3
d <- ggplot(pca.p.n, aes(PC2, PC3)) + 
  geom_point(aes(color = Year, shape = Year), size = 2) +
  xlab(paste0("PC2 (", signif(pve.p$pve[2], 3),"%)")) + 
  ylab(paste0("PC3 (", signif(pve.p$pve[3], 3),"%)")) +
  ggtitle("Parasite") +
  theme_bw() +
  theme(legend.position = "top",
        legend.title = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())

# join the plots
ggarrange(a, b, c, d,
          labels = c("A", "B", "C", "D"),
          ncol = 2, nrow = 2)

# join with patchwork (better)
combined_plot <- (a + b) / (c + d) & theme(legend.position = "bottom") 
combined_plot + plot_layout(guides = "collect")

combined_plot

################
## Tajima's D ##
################

## HOST
# read in the file of 1kb window size
taj.h <- read_table("parus1_swga_popgen/host/host_1kb.Tajima.D", 
                    col_names = TRUE) %>%
  filter(!N_SNPS == 0) %>%
  mutate(TajimaD = as.numeric(TajimaD))

# plot
ggplot(taj.h, aes(x = TajimaD)) +
   geom_histogram(binwidth = 1, closed = "left", 
                  boundary = 0, color = "white", 
                  fill = "black") +
   xlab("Tajima's D") +
   scale_y_continuous("Number of 1kb windows") +
   theme_bw()

# mean and se
mean(taj.h$TajimaD)
sd(taj.h$TajimaD)/sqrt(length(taj.h$TajimaD))

## PATH
# read in the file of 1kb window size
taj.p <- read_table("parus1_swga_popgen/pathogen/path_1kb.Tajima.D", 
                    col_names = TRUE) %>%
  filter(!N_SNPS == 0) %>%
  mutate(TajimaD = as.numeric(TajimaD))

# plot
ggplot(taj.p, aes(x = TajimaD)) +
   geom_histogram(binwidth = 1, closed = "left", 
                  boundary = 0, color = "white", 
                  fill = "black") +
   xlab("Tajima's D") +
   scale_y_continuous("Number of 1kb windows") +
   theme_bw()

# mean and se
mean(taj.p$TajimaD)
sd(taj.p$TajimaD)/sqrt(length(taj.p$TajimaD))

```

